Added SSL support to HttpConnector and accompanying functional test

0xPaul opened the pull request.

Commit:
Added SSL support to HttpConnector and accompanying functional test.
Added SslOptions and accompanying test.

a-barzanti commented on src/Net/Http/HttpOptions.php:
```diff
@@ -18,6 +24,14 @@ public function __construct()
+     */
+    public function getSslOptions()
+    {
+        return $this->sslOptions ?: $this->sslOptions = new SslOptions;
```
Not really sure abut this one, you expect the getter to give you the current value of the internal property, but in this case the property is null and the getter returns an object, maybe another name?

0xPaul replied:
Although not strictly a getter this is just a lazy loading pattern which is an internal optimization the caller does not need to be aware of. We could rename the method *getOrCreateSslOptions*, but again, the caller doesn't need to be aware of this implementation detail.

If we later remove the optimization and always create `SslOptions` in the constructor this would require another name change and yet, from the caller's perspective, nothing is different. Ergo the current name is correct because we should not leak minor implementation details to clients.

a-barzanti commented on src/Net/Http/HttpOptions.php:
```diff
@@ -18,6 +24,14 @@ public function __construct()
+     */
+    public function getSslOptions()
+    {
+        return $this->sslOptions ?: $this->sslOptions = new SslOptions;
```
Actually I didn't notice the assignment, still a little overpowered getter, I still suggest another name

markchalloner reviewed (commented):
TODO review

markchalloner commented on test/Unit/Porter/Net/Http/HttpOptionsTest.php:
```diff
@@ -62,7 +62,7 @@ public function testReplaceHeaders()
     public function testProxy()
     {
-        self::assertSame($host = 'https://foo.com:80', (new HttpOptions)->setProxy($host)->getProxy());
+        self::assertSame($host = 'https://example.com:80', (new HttpOptions)->setProxy($host)->getProxy());
```
Doesn't really matter as this is just a set/get test, but having both http*_s_* and port 80 seems odd.

markchalloner commented on test/Functional/Porter/Net/Http/HttpConnectorTest.php:
```diff
@@ -95,10 +103,86 @@ private function stopServer(Process $server)
+        $connector = new HttpConnector($options = new HttpOptions);
+        $options->getSslOptions()
+            ->setVerifyPeer(false)
+            ->setVerifyPeerName(false)
```
As discussed offline, it might be possible to avoid having to disable SSL verification. This would (probably) involve:

- adding an [OpenSSL root CA generation step](https://datacenteroverlords.com/2012/03/01/creating-your-own-ssl-certificate-authority/),
- signing the user certificate with the root CA key and add the `self::SSL_HOST` ip as a subject and
- passing the root CA certificate filename to `SslOptions::setCertificateAuthorityFilePath`.

Commit:
Changed HttpConnectorTest to use peer verification for SSL tests.
Added try/finally blocks to HttpConnectorTest to faciliatate proper cleanup to prevent stalled builds and knock-on test failures.

0xPaul commented:
Thanks @markchalloner for your review. Your suggestion to do proper peer verification has been implemented by passing the certificate to the connector. Normally this is all that would be needed, but due to a bug with peer name verification and IPv6 we had to specify the name manually; it works without this over IPv4, however.
